### DISCUSSION TITLE
Debugging with SSL on localhost using Android

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#381] [Q&A] [answered]
@adrianhall , I'm trying to setup an environment which will be easier to debug.

I have the ASP Core server running locally and the TodoApp running on an Android emulator. Now from what I understand, on Android you must use 10.0.2.2 to access the "localhost", however the issue is I can't use SSL when using that IP as it throws an exception

![image](https://user-images.githubusercontent.com/67002466/173203949-2dab20f0-3cce-4df6-88cb-81a53535c7d1.png)
 
and if I try to connect without SSL I get another exception stating I must use SSL.

![image](https://user-images.githubusercontent.com/67002466/173203989-c006b052-c579-41ad-9feb-eafb2527290f.png)

It works if I publish the server to Azure, but its harder to debug that way.

What's the recommended way, I'm losing so much time just trying to get things setup.

**Comment by fileman** [comment, marked as answer]
Hi,
you can use SSL on Android emulator, but you have to create a DelegatingHandler like this:
```
    public class AndroidEmulatorHandler : DelegatingHandler
    {
        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            HttpMessageHandler handler = this.InnerHandler;

            while (handler is DelegatingHandler)
            {
                handler = ((DelegatingHandler)handler).InnerHandler;
            }
#if DEBUG
            if (handler is HttpClientHandler httpClientHandler
                && httpClientHandler.ServerCertificateCustomValidationCallback == null)
            {
                httpClientHandler
                    .ServerCertificateCustomValidationCallback =
                        (message, cert, chain, errors) => true;
            }
#endif
            // Now do the request
            var response = await base.SendAsync(request, cancellationToken);

            return response;
        }
    }
```
and then add on DatasyncClientOptions:
```
var options = new DatasyncClientOptions
                {
                    HttpPipeline = new DelegatingHandler[] { new AndroidEmulatorHandler() }
                };
```

**Reply by henda79** [reply]
@fileman , thank you so much for posting the example. I was on the right track with the HttpPipeline but couldn't work it out.

While we are on with the DatasyncClientOptions, do you know the use of IdGenerator, its not documented (as far as I can see)

Thanks again !

**Reply by fileman** [reply]
with IdGenerator you can customize how the Id are generated, as example I'm using sequential guid generator, so I have:
```
        /// <summary>
        /// Used to generate custom Id.
        /// </summary>
        private readonly GuidGenerator GuidGenerator = new(); 

...
                var options = new DatasyncClientOptions
                {
                    OfflineStore = store,
                    HttpPipeline = new DelegatingHandler[] { new AndroidEmulatorHandler() },
                    IdGenerator = MyIdGenerator
                };
...

        /// <summary>
        /// Custom Id Generator
        /// </summary>
        /// <param name="tableName"></param>
        /// <returns></returns>
        protected string MyIdGenerator(string tableName = null)
        {
            // user code for custom id generation
            return GuidGenerator.Create().ToString();
        }
```
where GuidGenerator is something like [this](https://github.com/jhtodd/SequentialGuid/blob/master/SequentialGuid/Classes/SequentialGuid.cs)

You can use the table name parameter to restrict or differentiate the generation.
HTH

**Reply by henda79** [reply]
Thanks, I was asking as our old system I am porting from, allocated a range of int Id's to each device as we need to allow for easy human lookup in a table and a Guid is not easy.

The solution I need is to know the Id the client is using before remote insert, as the client may not always have a connection.

**Reply by fileman** [reply]
in "MyIdGenerator" you can call a remote api to get the int id range to use when offline and store and handle on client

**Reply by adrianhall** [reply]
No you can't - you would need the IdGenerator to be async for that.

**Reply by henda79** [reply]
@adrianhall, would you mind to elaborate on that ? I doesn't seem possible to mark the method as Async.

**Reply by adrianhall** [reply]
Correct - it isn't right now.  That means IdGenerator is not suitable for retrieving a range via a HTTP call.  (which was my point)

**Reply by fileman** [reply]
Directly no, but MyIdGenerator can handle the int range fetched somewhere with HttpClient and stored in local storage (or in another another datasync table?🤔)

**Reply by henda79** [reply]
I see there are plans for the docs, but are there plans to change IdGenerator ?

**Reply by adrianhall** [reply]
Feel free to file an issue with your suggestion on how to change IdGenerator.

**Reply by henda79** [reply]
I've not played with IdGenerator much, maybe @fileman would be more suited to this ? Maybe its just the case to already have a range ID's for that device as in my use case I would be using these offline so an async Http call wouldn't be needed, it would just call to the Db or similar local storage.

**Reply by ghost** [reply]
I tried adding the AndroidEmulatorHandler class by @fileman . And I was able to get around the "endpoint" must use secure error

An error I have now in .net maui is this:

![image](https://user-images.githubusercontent.com/81643512/175816070-561de4ea-7a67-45da-876f-afc376c982ec.png)


This doesn't occur in Xamarin.Forms, does anyone have a work around for this error?

**Reply by henda79** [reply]
I can't help with the MAUI side of things as I'm staying clear of that for a while, but another option would be to use a tool like NGROK

https://ngrok.com/
